Fix/websocket drops keeping bolts

leonardoaramaki opened the pull request:
<!-- INSTRUCTION: Keep the line below to notify all core developers about this new PR -->
@RocketChat/android

<!-- INSTRUCTION: Inform the issue number that this PR closes, or remove the line below -->

<!-- INSTRUCTION: Tell us more about your PR with screen shots if you can -->
WIP: Did some work on the network layer without entirely rewriting it, that being the fastest solution as possible on short time. These changes are supposed to handle websocket connection drops due to network intermittences such as changing from wi-fi to mobile network or total internet connection drops. The present PR is not supposed to be a permanent solution altogether and probably does not handle all edge cases.

leonardoaramaki committed:
Band-aid fix for the network issues without removing bolts, yet

leonardoaramaki committed:
Merge branch 'develop' into fix/websocket-drops-keeping-bolts

leonardoaramaki committed:
Removed forgotten unused imports

luciofm commented on android-ddp/src/main/java/chat/rocket/android_ddp/DDPClientImpl.java:
```diff
@@ -369,6 +378,28 @@ public void unsubscribeBaseListeners() {
+            .cast(RxWebSocketCallback.Failure.class)
+            .subscribe(
+                    task::setResult,
+                    err -> {
```
When `onSuccess` or `onError` have more than one line, extract it to a private method with a meaningful name, increases readability and make clear for other devs the intention.
Ex
```
.subscribe(task::setResult, this::setTaskError(task, throwable));

private void setTaskErro(TaskCompletionSource<RxWebSocketCallback.Failure> task, Throwable throwable) {
    if (err instanceof Exception) {
        task.setError((Exception) err);
    } else {
        task.setError(new Exception(err));
    }
}
```

luciofm commented on android-ddp/src/main/java/chat/rocket/android_ddp/DDPClientImpl.java:
```diff
@@ -383,7 +414,8 @@ private boolean sendMessage(String msg, @Nullable JSONBuilder json) {
                            TaskCompletionSource<?> taskForSetError) {
     if (!sendMessage(msg, json)) {
-      taskForSetError.trySetError(new DDPClientCallback.Closed(client));
+      if (taskForSetError != null)
```
Disclaimer: In the next few days I'll be creating an Coding Style/Guideline doc, and a review guidelines to keep and improve our code quality

Single `if` statements must have braces `{ }` or have the body in the same line.


```
// preferred style
if (condition) {
    body();
}

// also acceptable if fits right in one line
if (condition) body();

// not acceptable
if (condition)
    body();
```

luciofm commented on app/src/main/java/chat/rocket/android/activity/MainPresenter.java:
```diff
@@ -3,22 +3,24 @@
 import android.support.annotation.NonNull;
 import android.support.v4.util.Pair;
 
-import io.reactivex.Flowable;
```
I'll be creating a default Android Studio config, so we can avoid this import reordering on PRs.

luciofm commented on app/src/main/java/chat/rocket/android/activity/MainPresenter.java:
```diff
@@ -159,6 +161,19 @@ private void subscribeToSession() {
+                        view.showConnecting();
+                      }
+                    },
+                    err -> {
```
Don't ignore exceptions, do at least a `err.printStackTrace();` or if it is really safe to ignore it, add a comment explaining why.

Ignoring exceptions can make it harder to understand why something is not working, and printing the stacktrace make it obvious where the problem is,

leonardoaramaki replied:
I agree. I don't have intention to keep it this way. I agree it's bad and the actual codebase is ignoring exceptions wide spread. Just kept this way for the sake of rushing to release this PR for testing. The code needs much refining

luciofm commented on app/src/main/java/chat/rocket/android/activity/MainPresenter.java:
```diff
@@ -159,6 +161,19 @@ private void subscribeToSession() {
     addSubscription(subscription);
+
+    addSubscription(
+            RxJavaInterop.toV2Observable(connectivityManagerApi.getServerConnectivityAsObservable())
```
Could you please create an Issue to update the code base to RxJava 2, and add a TODO comment pointing to the issue whenever we need to use the RxJavaInterop for our own code.

leonardoaramaki replied:
Done!

luciofm commented on app/src/main/java/chat/rocket/android/service/RocketChatWebSocketThread.java:
```diff
@@ -3,24 +3,24 @@
 import chat.rocket.android.helper.LogIfError;
 import chat.rocket.android.helper.TextUtils;
 import chat.rocket.android.log.RCLog;
-import chat.rocket.core.models.ServerInfo;
```
We should try to avoid reordering imports on commits that also adds or removes imports, it makes harder to see what actually changed.

luciofm commented on app/src/main/java/chat/rocket/android/service/RocketChatWebSocketThread.java:
```diff
@@ -238,6 +252,35 @@ public void run() {
 
+                task.getResult().client.getOnFailureCallback().onSuccess(_task -> {
+                  ddpClient = null;
+                  CompositeDisposable subscriptions = new CompositeDisposable();
```
Try to keep a consistent naming, these are disposables not subscriptions :smiley:

leonardoaramaki replied:
I'll change those :) Just kept this way to keep the same naming convention already being used

luciofm commented on app/src/main/java/chat/rocket/android/service/RocketChatWebSocketThread.java:
```diff
@@ -238,6 +252,35 @@ public void run() {
+                  connectivityManager.notifyConnecting(hostname);
+                  subscriptions.add(
+                          RxJavaInterop.toV2Single(connect().retry())
+                                  .subscribe(
```
Extract to a method with meaningful name.

luciofm commented on app/src/main/java/chat/rocket/android/service/RocketChatWebSocketThread.java:
```diff
@@ -238,6 +252,35 @@ public void run() {
+                                            }
+                                          },
+                                          err -> {
+                                            subscriptions.dispose();
```
Don't ignore the exception. Is it safe to assume that no more subscriptions will be added to the CompositeDisposable after this?

luciofm commented on app/src/main/java/chat/rocket/android/service/observer/MethodCallObserver.java:
```diff
@@ -117,10 +118,10 @@ public void onUpdateResults(List<MethodCall> results) {
           } else if (exception instanceof DDPClientCallback.RPC.Timeout) {
             // temp "fix"- we need to rewrite the connection layer a bit
             errMessage = "{\"message\": \"Connection Timeout\"}";
-          } else if (exception instanceof RxWebSocketCallback.Failure) {
```
If it is not being used, just remove it... If it is something temporary, add a comment indicating that.

leonardoaramaki replied:
Ok then, I thought over that one and removed the commented code. :+1:

leonardoaramaki committed:
Upon failure, do unregister of current listeners and register them
all again with the new dppclient just reconnected

leonardoaramaki committed:
Handling timeout errors from service keep alive as a Failure callback event instead of a close one

leonardoaramaki committed:
Bumped back timeout for rpc method calls to 20000 millis

leonardoaramaki committed:
Closing response body from websocket frames if any to avoid resource leaks

rafaelks committed:
Merge branch 'develop' into fix/websocket-drops-keeping-bolts

leonardoaramaki committed:
A change of network type will now show a snackbar when a failure occur which will trigger a websocket reconnection flow

rafaelks committed:
Merge branch 'develop' into fix/websocket-drops-keeping-bolts

leonardoaramaki committed:
Using plurals for quantity strings

leonardoaramaki committed:
Added a semaphore lock in RocketChatService method getStarted()
preventing multiple threads inconsistently creating multiple websocket
threads for the same host and thus avoiding connecting to server multiple times unnecessarily.
Changed network changing fallback implemention to try to reconnect to
automatically and force-invalidating the session token before doing
this.

leonardoaramaki committed:
Removed unused imports and RxWebSocketCallback Failure class

leonardoaramaki committed:
Change circleci gradle opts

leonardoaramaki committed:
Change circleci gradle opts

rafaelks committed:
Increment build number to send another beta

philipbrito reviewed (commented):
Why calling `FLog.setMinimumLoggingLevel(...)` [two times](https://github.com/RocketChat/Rocket.Chat.Android/blob/40a217f0e57bb8b094c328fe3b56bf45d94a0f7e/rocket-chat-android-widgets/src/main/java/chat/rocket/android/widget/RocketChatWidgets.java)? And why setting `FLog.ERROR` instead of `FLog.VERBOSE` like recommended by [the documentation](http://frescolib.org/docs/troubleshooting.html)? Any special case? Anyway, take a look on [Fresco/ShowcaseApplication](https://github.com/facebook/fresco/blob/master/samples/showcase/src/main/java/com/facebook/fresco/samples/showcase/ShowcaseApplication.java) for reference.

leonardoaramaki commented:
@filipedelimabrito I didn't notice it was duplicated but I can rollback this one. I set to Flog.ERROR because Flog.VERBOSE spams a hell lot the logs with mostly useless information.

philipbrito commented:
The `Flog.VERBOSE` is especially useful when looking which `controllers` are been used with the `drawee`. This is a special cases to analyze, for example, how users avatars are shown in a RecyclerView, allowing a better/deep analyses.

leonardoaramaki commented:
Oh, I see. I'm sorry @filipedelimabrito. I never used Fresco before, didn't know it :)
I'll rollback on my next push!

leonardoaramaki committed:
Set back Fresco logger to verbose

leonardoaramaki committed:
Add a notification channel when target is api level >= 26

leonardoaramaki committed:
Removed unused imports

leonardoaramaki committed:
Merge branch 'fix/websocket-drops-keeping-bolts' of github.com:RocketChat/Rocket.Chat.Android into fix/websocket-drops-keeping-bolts

luciofm reviewed (approved):
Approving it..

Guys, lets try to keep PRs smaller and self contained, this one had a lot of stuff that was not related do the issue, like notifications channel (should be a new PR), CI fixes (another PR), etc...
